# build pipeline

In [1]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer

from sklearn.linear_model import LinearRegression
from sklearn.impute import SimpleImputer
from sklearn.impute import KNNImputer

from sklearn.preprocessing import StandardScaler

import joblib

import pandas as pd

imputer_median = SimpleImputer(strategy='median')
imputer_mean = SimpleImputer(strategy='mean')
knn_imputer = KNNImputer()

scaler = StandardScaler()

lr = LinearRegression()

ct = ColumnTransformer([
    ("imputer_median", imputer_median, ['CRIM', 'ZN', 'NOX', 'DIS', 'PTRATIO','B']),
    ("imputer_mean", imputer_mean, ['RM', 'LSTAT']),
    ("knn_imputer", knn_imputer, ['INDUS', 'CHAS','AGE', 'RAD', 'TAX'])
])

pipe = Pipeline(steps=[
    ("fill_na", ct),
    ("scale", scaler),
    ("moodel", lr)
])


df_train = pd.read_csv('train.csv')

X_train = df_train.drop(columns='MEDV')
y_train = df_train['MEDV']

pipe.fit(X_train, y_train)

pipe.score(X_train, y_train)

0.6379836760633126

In [2]:
pipe

Pipeline(steps=[('fill_na',
                 ColumnTransformer(transformers=[('imputer_median',
                                                  SimpleImputer(strategy='median'),
                                                  ['CRIM', 'ZN', 'NOX', 'DIS',
                                                   'PTRATIO', 'B']),
                                                 ('imputer_mean',
                                                  SimpleImputer(),
                                                  ['RM', 'LSTAT']),
                                                 ('knn_imputer', KNNImputer(),
                                                  ['INDUS', 'CHAS', 'AGE',
                                                   'RAD', 'TAX'])])),
                ('scale', StandardScaler()), ('moodel', LinearRegression())])

In [3]:
# save pipeline
joblib.dump(pipe, 'pipeline_2.pkl')

['pipeline_2.pkl']

TP - Métricas - Regresión Lineal

- R2 Train: 0.6379836760633126
- RMSE Train: 5.810058588033976

# load pipeline

In [4]:
import joblib

pipeline = joblib.load('pipeline_2.pkl')

In [5]:
pipeline

Pipeline(steps=[('fill_na',
                 ColumnTransformer(transformers=[('imputer_median',
                                                  SimpleImputer(strategy='median'),
                                                  ['CRIM', 'ZN', 'NOX', 'DIS',
                                                   'PTRATIO', 'B']),
                                                 ('imputer_mean',
                                                  SimpleImputer(),
                                                  ['RM', 'LSTAT']),
                                                 ('knn_imputer', KNNImputer(),
                                                  ['INDUS', 'CHAS', 'AGE',
                                                   'RAD', 'TAX'])])),
                ('scale', StandardScaler()), ('moodel', LinearRegression())])

In [6]:
import pandas as pd

df = pd.read_csv('test.csv')
df.head(1)

,CRIM,ZN,INDUS,CHAS,NOX,RM,AGE,DIS,RAD,TAX,PTRATIO,B,LSTAT,MEDV
0,0.35809,0.0,6.2,1.0,0.507,6.951,88.5,2.8617,8.0,307.0,17.4,391.7,9.71,26.7


In [7]:
X = df.drop(columns='MEDV')
y = df['MEDV']

pipeline.score(X, y)

0.5454570100629383

In [14]:
!apt-get install snap
!snap install docker

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
snap is already the newest version (2013-11-29-11).
0 upgraded, 0 newly installed, 0 to remove and 41 not upgraded.
/bin/bash: line 1: snap: command not found


# docker

In [12]:
# build image
!docker build -t inference-python-test .

/bin/bash: line 1: docker: command not found


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
# run container
!docker run -it --rm --name inference-python-test -v ./files:/files  inference-python-test

the input device is not a TTY.  If you are using mintty, try prefixing the command with 'winpty'


In [ ]:
### en Windows es probable que sea necesario poner el path absoluto para montar el volumen
docker run -it --rm --name inference-python-test -v "C:\AA1\Unidad6_MLOps\mlops-docker-example\files:/files" inference-python-test


# check docker output

In [ ]:
df_output = pd.read_csv('./files/output.csv')
df_output.head(1)

,MEDV_predicted
0,34.561835


In [ ]:
from sklearn.metrics import r2_score, root_mean_squared_error

y_pred = df_output.MEDV_predicted

round(r2_score(y, y_pred),3), round(root_mean_squared_error(y, y_pred), 3)

(0.545, 5.805)

TP - Métricas - Regresión Lineal

- R2 Test: 0.545
- RMSE Test: 5.805